stages/prompt: add ability to set separate labels and values for choices

ErikAhlund opened the pull request:
<!--
👋 Hi there! Welcome.

Please check the Contributing guidelines: https://docs.goauthentik.io/docs/developer-docs/#how-can-i-contribute
-->

## Details

<!--
Explain what this PR changes, what the rationale behind the change is, if any new requirements are introduced or any breaking changes caused by this PR.

Ideally also link an Issue for context that this PR will close using `closes #`
-->

Fixes #5197

This PR adds support for custom label and values in the select and radio components of the PromptStage by modifying the select object to accept an array of objects with value and label properties, while maintaining backward compatibility with string arrays. Key changes include:

Updated PromptStage.ts to handle both array of strings and array of { value: string, label: string } for select and radio inputs.

Added unit tests in authentik/stages/prompt/tests.py to verify new functionality and edge cases. (Further tests might be needed)

Updated relevant documentation to reflect the new option format in website/docs/add-secure-apps/flows-stages/stages/prompt/index.md

Please let me know if additional changes or tests are needed!

---

## Checklist

-   [ x ] Local tests pass (`ak test authentik/`)
-   [ x ] The code has been formatted (`make lint-fix`)

If an API change has been made

-   [ x ] The API schema has been updated (`make gen-build`)

If changes to the frontend have been made

-   [ x ] The code has been formatted (`make web`)

If applicable

-   [ x ] The documentation has been updated
-   [ x ] The documentation has been formatted (`make docs`)

ErikAhlund committed:
Choices can have value and label

Value and label are set turned to a string.
This will make choices into a Array<{ [key: string]: any; }> | null which at some point it should be a defined interface.

ErikAhlund committed:
Auto Updated schema.yml

ErikAhlund committed:
Used label and value in choice

Label and Value are used for Dropdown and RadioButtonGroup. Strings are still accepted.

ErikAhlund committed:
docs: Updated stages/prompt documentation

ErikAhlund committed:
Updated docs for initial-value

Also fixed wrong choice example in previous docs changes

ErikAhlund committed:
Check if choice is dict

Choice can of course be anything, we shouldn't assume it's string or dict

ErikAhlund committed:
Check if choice is dict for initial value

Same as before, choice can be anything. We check if it's explicitely a dict

ErikAhlund committed:
Added tests for dict choices

ErikAhlund committed:
ran make lint-fix

macmoritz commented on website/docs/add-secure-apps/flows-stages/stages/prompt/index.md:
```diff
@@ -65,7 +65,9 @@ A field placeholder, shown within the input field.
-In the case of `Radio Button Group` and `Dropdown` prompts, this field defines all possible values (choices). When interpreted as-is, only one value will be allowed (the placeholder string). When interpreted as expression, a list of values can be returned to define multiple choices. For example, `return ["first option", 42, "another option"]` defines 3 possible values.
+In the case of `Radio Button Group` and `Dropdown` prompts, this field defines all possible values (choices). When interpreted as-is, only one value will be allowed (the placeholder string). When interpreted as expression, a list of values can be returned to define multiple choices. For example, `return ["first option", 42, {"label": "another option", "value": "some value"}]` defines 3 possible v ...
+
+A choice can be represented as a string or other primitive (which will be turned into a string), or as an object with parameters `value` and/or `label`.Sow `return ["Option 1"]` is equivalent to `return [{"label": "Option 1", "value": "Option 1"}]`.
```
just a minor typo:  
```suggestion
A choice can be represented as a string or other primitive (which will be turned into a string), or as an object with parameters `value` and/or `label`. So `return ["Option 1"]` is equivalent to `return [{"label": "Option 1", "value": "Option 1"}]`.
```

macmoritz commented on authentik/stages/prompt/stage.py:
```diff
@@ -227,6 +231,18 @@ def get_prompt_challenge_fields(self, fields: list[Prompt], context: dict, dry_r
+                    {"value": str(choice.get("value", "")), "label": str(choice.get("label", ""))}
+                )
+            else:
+                clean.append(str(choice))  # Should we force it to be a dict with value + label?
```
IMHO parsing the list to a dict with value und label is cleaner, due to less code and complexity in the frontend.  
But I am not part of authentik, this is just my personal opinion.

ErikAhlund replied:
I agree, that way the data structure received by the frontend is consistent

ErikAhlund replied:
I just applied the changes to parse the list to a dict. I also made a new PromptChoiceSerializer which generates a PromptChoice type for the frontend. Currently it has a mandatory value and label (strings and other primitives are turned into this object by setting the value and label to the primitive)

ErikAhlund committed:
Apply typo fix from code review

Co-authored-by: macmoritz <49832924+macmoritz@users.noreply.github.com>
Signed-off-by: Erik Ahlund <erik@ahlund.me>

ErikAhlund committed:
stages/prompt: add PromptChoiceSerializer

choices are now a list of PromptChoiceSerializer instead of a generic DictField.
The PromptChallenge also forces the use of value/label object.

ErikAhlund committed:
web: use PromptChoice object

The front end can now safely use choices as an array of PromptChoice instead of it being either a string or an object.

ErikAhlund committed:
Merge branch 'feature/prompt-object-choices' of https://github.com/ErikAhlund/authentik into feature/prompt-object-choices

BeryJu committed:
slight revise

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

BeryJu committed:
Merge branch 'main' into feature/prompt-object-choices

BeryJu committed:
small ui

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

dominic-r commented on website/docs/add-secure-apps/flows-stages/stages/prompt/index.md:
```diff
@@ -65,7 +65,9 @@ A field placeholder, shown within the input field.
-In the case of `Radio Button Group` and `Dropdown` prompts, this field defines all possible values (choices). When interpreted as-is, only one value will be allowed (the placeholder string). When interpreted as expression, a list of values can be returned to define multiple choices. For example, `return ["first option", 42, "another option"]` defines 3 possible values.
+In the case of `Radio Button Group` and `Dropdown` prompts, this field defines all possible values (choices). When interpreted as-is, only one value will be allowed (the placeholder string). When interpreted as expression, a list of values can be returned to define multiple choices. For example, `return ["first option", 42, {"label": "another option", "value": "some value"}]` defines 3 possible v ...
+
+A choice can be represented as a string or other primitive (which will be turned into a string), or as an object with parameters `value` and/or `label`. So `return ["Option 1"]` is equivalent to `return [{"label": "Option 1", "value": "Option 1"}]`.
```
```suggestion
A choice can be a string (or any primitive that will be converted to a string) or an object with `value` and/or `label` properties. For example, `return ["Option 1"]` is equivalent to `return [{"label": "Option 1", "value": "Option 1"}]`.
```

dominic-r committed:
Apply suggestion from @dominic-r

Signed-off-by: Dominic R <dominic@sdko.org>

dominic-r committed:
Apply suggestion from @dominic-r

Signed-off-by: Dominic R <dominic@sdko.org>

dominic-r committed:
Apply suggestion from @dominic-r

Signed-off-by: Dominic R <dominic@sdko.org>

dominic-r reviewed (approved):
lgtm for docs
